pkg/appsec/appsec.go

```diff
@@ -0,0 +1,128 @@
+			w.Header().Add("Content-Type", "application/json")
+			w.WriteHeader(http.StatusNotImplemented)
+			if err := json.NewEncoder(w).Encode(reason); err != nil {
+				log.Error(err)
+			}
+		})
+	}
+	cfg, err := newConfig(coreconfig.Datadog)
+	if err != nil {
+		return disabled(fmt.Sprintf("appsec agent disabled due to a configuration error: %v", err)), errors.Wrap(err, "configuration: ")
+	}
+	if !cfg.Enabled {
+		log.Info("AppSec agent disabled. Set the environment variable `DD_APPSEC_ENABLED=true` or add the entry " +
+			"`appsec_config.enabled: true` to your datadog.yaml file")
+		return disabled("appsec agent disabled by configuration"), nil
+	}
+	return newIntakeReverseProxy(cfg.IntakeURL, cfg.APIKey, cfg.MaxPayloadSize, transport), nil
+}
+
+// newIntakeReverseProxy creates a reverse proxy to the intake backend using the
+// given transport round-tripper.
+// The reverse proxy handler also limits the request body size and adds extra
+// headers such as Dd-Api-Key and Via.
+func newIntakeReverseProxy(target *url.URL, apiKey string, maxPayloadSize int64, transport http.RoundTripper) http.Handler {
+	proxy := httputil.NewSingleHostReverseProxy(target)
+	via := fmt.Sprintf("trace-agent %s", info.Version)
+	// Wrap and overwrite the returned director to add extra headers
+	director := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		// Call the original director changing the request target
+		director(req)
+		// Set extra headers
+		req.Header.Set("Via", via)
+		req.Header.Set("Dd-Api-Key", apiKey)
+		if req.Body != nil && maxPayloadSize > 0 {
+			req.Body = apiutil.NewLimitedReader(req.Body, maxPayloadSize)
+		}
+	}
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
+		if err := json.NewEncoder(w).Encode(err); err != nil {
+			log.Error(err)
+		}
+	}
+	proxy.Transport = transport
+	proxy.ErrorLog = stdlog.New(logutil.NewThrottled(5, 10*time.Second), "Appsec backend proxy: ", 0)
+	return withMetrics(proxy)
+}
+
+const (
+	appSecRequestMetricsPrefix            = "datadog.trace_agent.appsec."
+	appSecRequestCountMetricsID           = appSecRequestMetricsPrefix + "request"
+	appSecRequestDurationMetricsID        = appSecRequestMetricsPrefix + "request_duration_ms"
+	appSecRequestPayloadSizeMetricsID     = appSecRequestMetricsPrefix + "request_payload_size"
+	appSecRequestPayloadTooLargeMetricsID = appSecRequestMetricsPrefix + "request_payload_too_large"
+)
+
+func withMetrics(proxy *httputil.ReverseProxy) http.Handler {
+	// Error metrics through the reverse proxy error handler
+	errorHandler := proxy.ErrorHandler
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		if err == apiutil.ErrLimitedReaderLimitReached {
+			if err := metrics.Count(appSecRequestPayloadTooLargeMetricsID, 1, metricsTags(req), 1); err != nil {
```

Comment:
Frankly we never check the error return value for this, anywhere 🙂

Reply:
Ok, so we need to adjust errcheck in .golangci file tell not to fail on these calls, like it's done for `log.Error` =)

Reply:
I changed `.golangci` so that `errcheck` doesn't trigger for calls to functions from `pkg/trace/metrics`.
And I simplified back the code by removing the error checks.